arenadata_toolkit: add column tablespace_location

red1452 opened the pull request:
New column tablespace_location was added to relations db_files_current,
__db_files_current, __db_files_current_unmapped and db_files_history. This field
contains physical path to a directory, that contains file in the file column
(but without db oids). For pg_default and pg_global this column is a subpath of
a full file path in file column. For user's tablespaces this column contains the
result of gp_tablespace_segment_location(oid) for current tablespace oid and
current segment id.

At update from version without tablespace_location to version with
tablespace_location current table db_files_history (if exists ) will be renamed
to db_files_history_backup_YYYYMMDDtHHMMSS, where YYYYMMDDtHHMMSS is timestamp
iso 8601 basic short. After update before using toolkit it should execute:
SELECT arenadata_toolkit.adb_create_tables();

The most tests were updated: command "DROP SCHEMA arenadata_toolkit cascade;"
was added to the end of tests. It needs to keep clear environment after each
tests.

New tests were added: tablespace_location and upgrade_test. Test
tablespace_location checks new field with different values: pg_default,
pg_global and custom tablespace. Test upgrade_test checks updates from all
previous versions to current.

Stolb27 committed:
Merge pull request #659 from arenadata/arenadata53

Patch set 53

red1452 committed:
arenadata_toolkit: add column tablespace_location (#662)

New column "tablespace_location" was added to relations "db_files_current",
"__db_files_current", "__db_files_current_unmapped" and "db_files_history".
This field contains physical path to a directory, that contains file in the
"file" column (but without db oids). For "pg_default" and "pg_global" this
column is a subpath of a full file path in file column. For user's tablespaces
this column contains the result of gp_tablespace_segment_location(oid) for
current tablespace oid and current segment id.

At update from version without "tablespace_location" to version with
"tablespace_location" current table "db_files_history" (if exists ) will be
renamed to "db_files_history_backup_YYYYMMDDtHHMMSS", where "YYYYMMDDtHHMMSS" is
timestamp iso 8601 basic short. After update before using toolkit it should
execute: "SELECT arenadata_toolkit.adb_create_tables();"

The most tests were updated: command "DROP SCHEMA arenadata_toolkit cascade;"
was added to the end of tests. It needs to keep clear environment after each
tests.

New tests were added: "tablespace_location" and "upgrade_test". Test
"tablespace_location" checks new field with different values: "pg_default",
"pg_global" and custom tablespace. Test "upgrade_test" checks updates from
version "1.0" to current.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60267

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879077

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879078

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60277

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879703

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879704

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879701

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879695

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/882259

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/882260

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/883008

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/883009

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/883010

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/883011

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,369 @@
+			ALTER TABLE arenadata_toolkit.db_files_current
+			ADD COLUMN tablespace_location TEXT;
+		END IF;
+	END IF;
```
Why not to use IF EXISTS .. **AND** NOT EXISTS .. ?
```
IF EXISTS (SELECT 1 FROM pg_tables
			   WHERE schemaname = 'arenadata_toolkit' AND tablename = 'db_files_current')
	AND NOT EXISTS (SELECT 1
				   FROM information_schema.columns
				   WHERE table_schema='arenadata_toolkit' AND
				         table_name='db_files_current' AND
				         column_name='tablespace_location')
	THEN
		ALTER TABLE arenadata_toolkit.db_files_current
		ADD COLUMN tablespace_location TEXT;
	END IF;
```

bimboterminator1 replied:
The remark is related to all IF THEN parts of this function.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,369 @@
+			ADD COLUMN tablespace_location TEXT;
+		END IF;
+	END IF;
+
```
In the first paragraph of specification:
"Column tablespace_location may already exists in such case: user has arenadata_toolkit 1.2, **then remove it, and it install again**."  What does it mean? It's for user, that he has to reinstall extension himself?

red1452 replied:
It means, that there are cases, when user can have these tables without installed extension (with or without column `tablespace_location`). And if table exists and there is column `tablespace_location` we should do nothing, if there is not this column, we should add this column to the table only.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,369 @@
+		END IF;
+	END IF;
+
+	EXECUTE FORMAT($fmt$ALTER TABLE IF EXISTS arenadata_toolkit.db_files_history
+						RENAME TO db_files_history_backup_%1$s;$fmt$,
+		to_char(now(), 'YYYYMMDD"t"HH24MISS'));
```
7th point of specification: "Rename arenadata_toolkit.db_files_history to arenadata_toolkit.db_files_history_backup_{{ timestamp }}  if it exists,"

In this code if the `arenadata_toolkit.db_files_history` does not exist, some NOTICE messages are written. Is it OK for us?

bimboterminator1 replied:
In other words, the NOTICE messages are inevitable in current context? Or we can avoid them?

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60556

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/894685

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/894686

red1452 committed:
fixes

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60562

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/894833

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/894834

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/894832

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,371 @@
+LANGUAGE plpgsql VOLATILE
+EXECUTE ON MASTER;
+
+SELECT arenadata_toolkit.check_and_alter_tablespace_location();
+-- This function is needed only once, for correct update tables and view
+DROP FUNCTION arenadata_toolkit.check_and_alter_tablespace_location();
```
Do we really need to create a separate function for the upgrade purposes (arenadata_toolkit--1.4.sql file now contains extra function and commands)?
```
DO $$ BEGIN
END$$
 LANGUAGE plpgsql;
```
How critical for us to use VOLATILE and EXECUTE ON MASTER options?

red1452 replied:
> How critical for us to use VOLATILE and EXECUTE ON MASTER options?

Now, there is not any difference, because `VOLATILE` and `EXECUTE ON MASTER` are default parameters of plpgsql functions. But I think, it is better to specify these values. It can help us in future, if default values will be changed.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,371 @@
+	IF EXISTS (SELECT 1 FROM pg_tables
+			   WHERE schemaname = 'arenadata_toolkit' AND
+			         tablename = 'db_files_history')
+	AND NOT EXISTS (SELECT 1
+					FROM information_schema.columns
+					WHERE table_schema='arenadata_toolkit' AND
+					      table_name='db_files_history' AND
+					      column_name='tablespace_location')
+	THEN
+		EXECUTE FORMAT($fmt$ALTER TABLE arenadata_toolkit.db_files_history
+							RENAME TO db_files_history_backup_%1$s;$fmt$,
+			to_char(now(), 'YYYYMMDD"t"HH24MISS'));
+	END IF;
```
"Rename arenadata_toolkit.db_files_history to arenadata_toolkit.db_files_history_backup_{{ timestamp }}  if it exists, 

We backup only if the `tablespace_location` does not exist?

red1452 replied:
This backup is needed for the next step:
```sql
INSERT INTO arenadata_toolkit.db_files_history 
SELECT
	oid,
	table_name,
	table_schema,
	type,
	storage,
	table_parent_table,
	table_parent_schema,
	table_database,
	table_tablespace,
	"content",
	segment_preferred_role,
	hostname,
	address,
	file,
	modifiedtime,
	file_size,
	NULL as tablespace_location,
	collecttime
FROM arenadata_toolkit.db_files_history_backup_{{ timestamp }};
```
Only for case, when there is not column `tablespace_location` at the table `arenadata_toolkit.db_files_history`.

red1452 committed:
upgrade srcipt and test optimizations

red1452 committed:
fix typo

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60614

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/897886

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/897887

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+		IF NOT EXISTS (SELECT 1
+					   FROM pg_attribute
+					   WHERE attname = 'tablespace_location' AND
+					         NOT attisdropped AND
```
What's the motivation behind using `NOT attisdropped `? If attribute is dropped we suppose that it exists anyway and there is no need to add this column? Is it expected behaviour?

bimboterminator1 commented on gpcontrib/arenadata_toolkit/expected/adb_vacuum_strategy_test.out:
```diff
@@ -1,10 +1,13 @@
  
 (1 row)
 
+RESET client_min_messages;
```
You decided to RESET here in order to leave vacuum tests untouched? (I see that vacuum tests have their own SET RESET pair).
Why in `adb_get_relfilenodes_test.sql` the whole test is just wrapped by SET at the begining and RESET at the end, but meanwhile  in `adb_collect_table_stats_test.sql`  `client_min_messages` is SET at the begining, and at the end it's not RESET? 
Maybe, if you've started to make the tests idependent and clear the enviroment, it would be also better to **unify** the style with warnings? Otherwise there is some inconsistency in style.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/sql/tablespace_location.sql:
```diff
@@ -0,0 +1,100 @@
+					 'gp_segment_configuration') -- example of table at pg_global
+ORDER BY table_name, content;
+
+SELECT
```
Why do we need 2 identical queries one after another?

red1452 replied:
It is the same queries for different relations. It needs to check each relation, where there is new column `tablespace_location`.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/sql/tablespace_location.sql:
```diff
@@ -0,0 +1,100 @@
+					 'gp_segment_configuration') -- example of table at pg_global
+ORDER BY table_name, content;
+
+SELECT arenadata_toolkit.adb_collect_table_stats();
```
May you add simple comments for what is being tested here?

red1452 replied:
This query is needed to add data to tables at arenadata_toolkit. This query does not check anything.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/sql/tablespace_location.sql:
```diff
@@ -0,0 +1,100 @@
+-- Change log level to disable notice messages from PL/pgSQL
+SET client_min_messages=WARNING;
+SELECT arenadata_toolkit.adb_create_tables();
+RESET client_min_messages;
```
May be supress NOTICE messages for the whole test, not locally?

red1452 replied:
I try to minimize supressing NOTICE messages. It may be important in case of errors.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/expected/adb_collect_table_stats_test.out:
```diff
@@ -317,3 +317,4 @@ ORDER BY oid;
 DROP TABLE part_table;
 DROP FUNCTION remove_partition_from_db_files_history();
 DROP EXTENSION arenadata_toolkit;
+DROP SCHEMA arenadata_toolkit CASCADE;
```
In that file  `client_min_messages` is SET at the begining, but is  at the end there is not RESET.

red1452 committed:
add missing RESET client_min_messages

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60685

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/902482

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/902483

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+
+DO $$
+BEGIN
+	IF EXISTS (SELECT 1
```
why not simple one select?
```sql
select c.oid is not null and attrelid is null
from pg_class c
join pg_namespace n on n.oid = relnamespace
left join pg_attribute a on attrelid = c.oid and attname = 'tablespace_location'
where relname = 'db_files_current' and nspname = 'arenadata_toolkit';
```

red1452 replied:
There are three cases:
1) There is not this table (nothing to change).
2) There is table with column `tablespace_location` (nothing to change).
3)  There is table without column `tablespace_location`. Only at this case we should add column `tablespace_location`.

RekGRpth replied:
pointed query returns
1) NULL
2) false
3) true

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+
+DO $$
+BEGIN
+	IF EXISTS (SELECT 1
```
why not simple one sql?
```sql
DROP TABLE if exists pg_temp.db_files_current;
```

red1452 replied:
As It was discussed with @bimboterminator1, At this case there is extra NOTICE message (if table does not exist).

bimboterminator1 replied:
If having NOTICes is OK, I don't mind about writing a more optimal command

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+
+DO $$
+BEGIN
+	IF EXISTS (SELECT 1
```
why we need all these conditions?
why not simple create or replace view?

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+					         attrelid = 'arenadata_toolkit.db_files_history'::regclass)
+		THEN
+			EXECUTE FORMAT($fmt$ALTER TABLE arenadata_toolkit.db_files_history
+								RENAME TO db_files_history_backup_%1$s;$fmt$,
```
```suggestion
								RENAME TO %1$I;$fmt$,
				to_char(now(), '"db_files_history_backup_"YYYYMMDD"T"HH24MISS'));
```

red1452 replied:
What's the point of this change? Only how does it look?

RekGRpth replied:
1) specification contains upper `T` between date and time
2) qualificator `I` corresponds to identifier

red1452 replied:
> 1. specification contains upper T between date and time

```
postgres=# alter table arenadata_toolkit.db_files_history rename to db_files_history_backup_20231221T141316
postgres-# ;
ALTER TABLE
postgres=# select relname from pg_class where relname like 'db_files_history_backup_%';
                          relname                           
------------------------------------------------------------
 db_files_history_backup_20231221t141316
 db_files_history_backup_20231221t141316_1_prt_p202312
 db_files_history_backup_20231221t141316_1_prt_default_part
```

RekGRpth replied:
> alter table arenadata_toolkit.db_files_history rename to db_files_history_backup_20231221T141316

try
```sql
alter table arenadata_toolkit.db_files_history rename to "db_files_history_backup_20231221T141316";
```
or use `I` qualificator with upper `T`

red1452 replied:
Done.

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+					ELSE (SELECT tblspc_loc
+						  FROM gp_tablespace_segment_location(t.oid)
+						  WHERE gp_segment_id = dbf.segindex)
+					END AS tablespace_location
```
may be move this column to arenadata_toolkit.__db_segment_files view, because of it already has join with gp_segment_configuration?

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,399 @@
+		)
+		WITH (appendonly = false) DISTRIBUTED RANDOMLY;
+
+		GRANT SELECT ON TABLE arenadata_toolkit.db_files_current TO public;
```
why this grant moved into if? in first version it was outside of if

red1452 replied:
If table exists before executing `arenadata_toolkit.adb_create_tables()`, we haven't change grants.

RekGRpth replied:
Why did this behavior only appear in version 1.4?

red1452 replied:
Because, only for version "1.4" there is renaming table `db_files_history` to `db_files_history_backup_...`. After update from one of previous versions to "1.4" user should execute  `adb_create_tables()` for creating new table `db_files_history`, other tables have already existed and this function have not change these tables (include grants).

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,155 @@
+LANGUAGE plpgsql VOLATILE
+EXECUTE ON MASTER;
+
+SELECT do_upgrade_test_for_arenadata_toolkit('1.0', '1.1') ORDER BY 1;
```
I think these tests are out of specification. As far as I understand, we need four tests, each of which first installs the previous version of the extension (without the new column), and at the end - the current version of the extension (with the new column), and performs some actions in between these installations. In addition to this, the last test requires to first install the extension from the bundle, i.e. the way it was before it became an extension.

red1452 replied:
I think another: 1-3 steps must check update between two adjacent versions of toolkit (for each pair). The last step must check installing toolkit at non-empty schema (if toolkit was installed and removed).

RekGRpth replied:
> pre "extension era" arendata_toolkit schema exists

I think it means arenadata_toolkit from bundle as it was before it became an extension.

RekGRpth replied:
How do tests with versions 1.0 -> 1.1 -> 1.2 -> 1.3 relate to the current patch to add a new column?

red1452 replied:
I think these tests should be added earlier, and if new version of toolkit will be added, this test must be updated (added new execution of `do_upgrade_test_for_arenadata_toolkit` for new version).

red1452 replied:
> I think it means arenadata_toolkit from bundle as it was before it became an extension.

I try to imitate it by executing `adb_create_tables()` and  `adb_collect_table_stats()` and dropping extension. Schema `areandata_toolkit` is not dropped and all tables exist.

RekGRpth replied:
> I think these tests should be added earlier

Still, I don’t quite understand what tests 1.0 -> 1.1 -> 1.2 -> 1.3 check?



> I try to imitate it by executing `adb_create_tables()` and `adb_collect_table_stats()` and dropping extension. Schema `areandata_toolkit` is not dropped and all tables exist.

OK, but specification says that we should call adb_create_tables and adb_collect_table_stats in this check.
Moreover, the bundle version of the extension contains objects with a different structure than those that remain after removing the extension.

RekGRpth replied:
I think we need tests for following
1) bundle version -> 1.4
2) 1.0 -> 1.4
3) 1.1 -> 1.4
4) 1.2 -> 1.4
5) 1.3 -> 1.4

with scenarios from specification.
Because the customers will have it.

RekGRpth replied:
Is it possible to organize a cycle for updating from all previous versions to the current version, so as not to edit these tests in new versions?

red1452 replied:
> Is it possible to organize a cycle for updating from all previous versions to the current version, so as not to edit these tests in new versions?

I know only how can we get first and last versions of extension (without intermediate versions):
```sql
postgres=# select version from pg_available_extension_versions where name = 'arenadata_toolkit';
 version 
---------
 1.4
 1.0
(2 rows)
```
But I little optimize the test, to do all at one query.

red1452 committed:
optimization of upgrade script

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60838

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/912335

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/912336

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/912333

red1452 committed:
fix check to remove table from pg_temp, optimize upgrade test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60867

BenderArenadata commented:
Failed job Orca unittests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/914402

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/914405

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/914406

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/914397

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/914399

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,379 @@
+BEGIN
+	IF EXISTS (SELECT 1
+			   FROM pg_tables
+			   WHERE schemaname LIKE 'pg_temp%' AND
```
Why did you change equality to start checking?

red1452 replied:
```sql
postgres=# select schemaname, tablename from pg_tables where schemaname LIKE 'pg_temp%' AND tablename = 'db_files_current';
 schemaname |    tablename     
------------+------------------
 pg_temp_25 | db_files_current
(1 row)
```

RekGRpth replied:
Why are you removing the db_files_current table from the pg_temp schema if it is in the pg_temp_25 schema?

red1452 replied:
```sql
postgres=#
CREATE TEMPORARY TABLE IF NOT EXISTS pg_temp.db_files_current
        (
                oid BIGINT,
                table_name TEXT,
                table_schema TEXT,
                type CHAR(1),
                storage CHAR(1),
                table_parent_table TEXT,
                table_parent_schema TEXT,
                table_database TEXT,
                table_tablespace TEXT,
                "content" INTEGER,
                segment_preferred_role CHAR(1),
                hostname TEXT,
                address TEXT,
                file TEXT,
                modifiedtime TIMESTAMP WITHOUT TIME ZONE,
                file_size BIGINT,
                tablespace_location TEXT
        )
        DISTRIBUTED RANDOMLY;
CREATE TABLE
postgres=#
select schemaname, tablename from pg_tables where schemaname LIKE 'pg_temp%' AND tablename = 'db_files_current';
 schemaname |    tablename     
------------+------------------
 pg_temp_25 | db_files_current
(1 row)

postgres=#
DROP TABLE pg_temp.db_files_current;
DROP TABLE
postgres=#
select schemaname, tablename from pg_tables where schemaname LIKE 'pg_temp%' AND tablename = 'db_files_current';
 schemaname | tablename 
------------+-----------
(0 rows)
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,379 @@
+				WHEN 'pg_default' = t.spcname THEN gpconf.datadir || '/base'
+				WHEN 'pg_global' = t.spcname THEN gpconf.datadir || '/global'
+				ELSE (SELECT tblspc_loc
+					  FROM gp_tablespace_segment_location(t.oid)
```
May we move this to arenadata_toolkit.__db_segment_files because it already has pg_tablespace and gp_segment_configuration?

red1452 replied:
At specification there is not requirement to change `arenadata_toolkit.__db_segment_files`.

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/tablespace_location.sql:
```diff
@@ -0,0 +1,100 @@
+		THEN '<SEGMENT_BASE_DIR>/global'
+		ELSE files.tablespace_location
+	END AS tablespace_location
+FROM arenadata_toolkit.db_files_history files
```
how about 3 in 1?
```suggestion
FROM (
	SELECT '__db_files_current' name, content, table_name, table_tablespace, tablespace_location, segment_preferred_role::CHAR(1) FROM arenadata_toolkit.__db_files_current UNION
	SELECT 'db_files_current' name, content, table_name, table_tablespace, tablespace_location, segment_preferred_role::CHAR(1) FROM arenadata_toolkit.db_files_current UNION
	SELECT 'db_files_history' name, content, table_name, table_tablespace, tablespace_location, segment_preferred_role::CHAR(1) FROM arenadata_toolkit.db_files_history
	) files
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+	LOOP
+		RETURN QUERY SELECT do_upgrade_test_for_arenadata_toolkit(
+								version.column1::TEXT,
+								(SELECT default_version
```
If we always update the extension to the latest version, then we don’t have to specify it explicitly by simply writing
```sql
ALTER EXTENSION arenadata_toolkit UPDATE;
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+
+-- Simple check: only create and alter extension:
+	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
+	IF (from_version != '1.0')
```
```suggestion
	IF from_version != '1.0'
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+						UPDATE TO %1$I;$fmt$, to_version);
+
+-- Check the result
+	IF to_version = (SELECT extversion
```
```suggestion
	IF to_version = extversion FROM pg_extension WHERE extname = 'arenadata_toolkit' THEN
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
+							UPDATE TO %1$I;$fmt$, from_version);
+	END IF;
+	EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
```
```suggestion
	ALTER EXTENSION arenadata_toolkit UPDATE;
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+-- which were added at version 1.4
+	PERFORM arenadata_toolkit.adb_create_tables();
+
+	IF EXISTS (SELECT 1
```
```suggestion
	IF 4 = (SELECT count(*) FROM information_schema.columns
	WHERE table_schema = 'arenadata_toolkit' AND
	table_name IN ('db_files_current', '__db_files_current', '__db_files_current_unmapped', 'db_files_history') AND
	column_name = 'tablespace_location')
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+
+-- Cleanup
+DROP FUNCTION do_test_for_each_versions();
+DROP FUNCTION do_upgrade_test_for_arenadata_toolkit(TEXT, TEXT);
```
I still don't see a test for the bundled version of the extension.

red1452 replied:
Bundle has three tables:
```
adb=# SELECT c.relname
FROM pg_class c
LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
WHERE n.nspname = 'arenadata_toolkit';
               relname               
-------------------------------------
 db_files_history
 db_files_history_1_prt_default_part
 db_files_history_1_prt_p202312
 daily_operation
 operation_exclude
(5 rows)
```
These tables is created at function `mock_manual_installation()` at test `arenadata_toolkit_test`. This test checks how last version of arenadata_toolkit will be installed to bundle.

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,175 @@
+BEGIN
+	FOR version IN
+		SELECT *
+		FROM (VALUES ('1.0'),('1.1'),('1.2'),('1.3'))
```
```suggestion
SELECT do_upgrade_test_for_arenadata_toolkit('1.0');
SELECT do_upgrade_test_for_arenadata_toolkit('1.1');
SELECT do_upgrade_test_for_arenadata_toolkit('1.2');
SELECT do_upgrade_test_for_arenadata_toolkit('1.3');
```

red1452 committed:
change "T" to "t" for backup, optimize tests

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60907

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/917277

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/917274

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/917267

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/917266

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/917888

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/918157

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/919339

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,375 @@
+BEGIN
+	IF (SELECT a.attrelid IS NULL
+		FROM pg_class c
+		LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
```
Why is there a LEFT JOIN here? Can there be a table without a schema? Moreover, the condition explicitly indicates the scheme and the LEFT JOIN loses its meaning.

red1452 committed:
optimize upgrade script

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60946

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/919476

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/919478

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/919473

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/919467

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/919469

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/920138

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,375 @@
+	IF (SELECT a.attrelid IS NULL
+		FROM pg_class c
+		JOIN pg_namespace n ON n.oid = c.relnamespace AND
+		                       n.nspname = 'arenadata_toolkit'
```
Typically, a condition, that is not related to both tables at the same time, is moved to the where block:
```sql
	IF (SELECT a.attrelid IS NULL
		FROM pg_class c
		JOIN pg_namespace n ON n.oid = c.relnamespace
		LEFT JOIN pg_attribute a ON a.attrelid = c.oid AND a.attname = 'tablespace_location'
		WHERE c.relname = 'db_files_current' AND n.nspname = 'arenadata_toolkit')
	THEN
```

red1452 replied:
Query plans of your and my query is the same:
```
postgres=# explain (costs off) SELECT a.attrelid IS NULL
                FROM pg_class c
                JOIN pg_namespace n ON n.oid = c.relnamespace
                LEFT JOIN pg_attribute a ON a.attrelid = c.oid AND a.attname = 'tablespace_location'
                WHERE c.relname = 'db_files_current' AND n.nspname = 'arenadata_toolkit'
postgres-# ;
                            QUERY PLAN                            
------------------------------------------------------------------
 Hash Join
   Hash Cond: (c.relnamespace = n.oid)
   ->  Hash Right Join
         Hash Cond: (a.attrelid = c.oid)
         ->  Seq Scan on pg_attribute a
               Filter: (attname = 'tablespace_location'::name)
         ->  Hash
               ->  Seq Scan on pg_class c
                     Filter: (relname = 'db_files_current'::name)
   ->  Hash
         ->  Seq Scan on pg_namespace n
               Filter: (nspname = 'arenadata_toolkit'::name)
 Optimizer: Postgres query optimizer
(13 rows)

postgres=# explain (costs off) SELECT a.attrelid IS NULL
                FROM pg_class c
                JOIN pg_namespace n ON n.oid = c.relnamespace AND
                                       n.nspname = 'arenadata_toolkit'
                LEFT JOIN pg_attribute a ON a.attrelid = c.oid AND
                                            a.attname = 'tablespace_location'
                WHERE c.relname = 'db_files_current'
postgres-# ;
                            QUERY PLAN                            
------------------------------------------------------------------
 Hash Join
   Hash Cond: (c.relnamespace = n.oid)
   ->  Hash Right Join
         Hash Cond: (a.attrelid = c.oid)
         ->  Seq Scan on pg_attribute a
               Filter: (attname = 'tablespace_location'::name)
         ->  Hash
               ->  Seq Scan on pg_class c
                     Filter: (relname = 'db_files_current'::name)
   ->  Hash
         ->  Seq Scan on pg_namespace n
               Filter: (nspname = 'arenadata_toolkit'::name)
 Optimizer: Postgres query optimizer
(13 rows)
```
As for me, my variant is more readable.

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,134 @@
+-- from "DROP SCHEMA arenadata_toolkit CASCADE;"
+SET client_min_messages=WARNING;
+
+-- We have only first and last versions of arenadata_toolkit and scripts to update
```
I don't think so, instead we have [all intermediate versions](https://github.com/arenadata/gpdb/blob/e73000417c345a3e6ec913c17fb9eb04d8e6f8ee/gpcontrib/arenadata_toolkit/Makefile#L16) of update scripts, so, you may simplify upgrade tests.

red1452 replied:
You can check it after clean build.
```
evgeniy@evgeniy-pc:~/gpdb/adb-6.x-gpdb$ ls -la ../build/gpdb_adb_6/share/postgresql/extension/ | grep arenadata
-rw-r--r-- 1 evgeniy evgeniy  1728 Dec 23 21:52 arenadata_toolkit--1.0--1.1.sql
-rw-r--r-- 1 evgeniy evgeniy 11403 Dec 23 21:52 arenadata_toolkit--1.0.sql
-rw-r--r-- 1 evgeniy evgeniy  3700 Dec 23 21:52 arenadata_toolkit--1.1--1.2.sql
-rw-r--r-- 1 evgeniy evgeniy  2573 Dec 23 21:52 arenadata_toolkit--1.2--1.3.sql
-rw-r--r-- 1 evgeniy evgeniy 10944 Dec 23 21:52 arenadata_toolkit--1.3--1.4.sql
-rw-r--r-- 1 evgeniy evgeniy 30348 Dec 23 21:52 arenadata_toolkit--1.4.sql
-rw-r--r-- 1 evgeniy evgeniy   200 Dec 23 21:52 arenadata_toolkit.control
```
This [line](https://github.com/arenadata/gpdb/blob/e73000417c345a3e6ec913c17fb9eb04d8e6f8ee/gpcontrib/arenadata_toolkit/Makefile#L16) generates only the last version of extension.
```
postgres=# create extension arenadata_toolkit version '1.0';
CREATE EXTENSION
postgres=# drop extension arenadata_toolkit ;
DROP EXTENSION
postgres=# create extension arenadata_toolkit version '1.1';
ERROR:  could not stat file "/home/evgeniy/gpdb/build/gpdb_adb_6/share/postgresql/extension/arenadata_toolkit--1.1.sql": No such file or directory
postgres=# create extension arenadata_toolkit version '1.2';
ERROR:  could not stat file "/home/evgeniy/gpdb/build/gpdb_adb_6/share/postgresql/extension/arenadata_toolkit--1.2.sql": No such file or directory
postgres=# create extension arenadata_toolkit version '1.3';
ERROR:  could not stat file "/home/evgeniy/gpdb/build/gpdb_adb_6/share/postgresql/extension/arenadata_toolkit--1.3.sql": No such file or directory
postgres=# create extension arenadata_toolkit version '1.4';
CREATE EXTENSION
postgres=# drop extension arenadata_toolkit ;
DROP EXTENSION
```

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,375 @@
+BEGIN
+	IF EXISTS (SELECT 1
+			   FROM pg_tables
+			   WHERE schemaname LIKE 'pg_temp%' AND
```
```suggestion
			   WHERE schemaname = 'pg_temp_'||(select sess_id from pg_stat_activity where pid = pg_backend_pid()) AND
```

red1452 replied:
Are you sure, that `pg_backend_pid()` will be equal to pid of backend, which created table at pg_temp?

RekGRpth commented:
OK, I'm waiting tests on CI, locally they are passed.

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/921274

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/921856

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/922679

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/923782

red1452 committed:
empty commit to rerun tests

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61168

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932620

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932609

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932610

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932621

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/933826

red1452 committed:
Remove command, which clear directory for custom tablespace

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61220

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/935315

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/935316

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/935308

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/935307

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/935309

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/935310

red1452 committed:
add drop temp table

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/952260

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/952261

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/952251

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/952252

red1452 committed:
arenadata_toolkit: add column tablespace_location (#662)

New column "tablespace_location" was added to relations "db_files_current",
"__db_files_current", "__db_files_current_unmapped" and "db_files_history".
This field contains physical path to a directory, that contains file in the
"file" column (but without db oids). For "pg_default" and "pg_global" this
column is a subpath of a full file path in file column. For user's tablespaces
this column contains the result of gp_tablespace_segment_location(oid) for
current tablespace oid and current segment id.

At update from version without "tablespace_location" to version with
"tablespace_location" current table "db_files_history" (if exists ) will be
renamed to "db_files_history_backup_YYYYMMDDtHHMMSS", where "YYYYMMDDtHHMMSS" is
timestamp iso 8601 basic short. After update before using toolkit it should
execute: "SELECT arenadata_toolkit.adb_create_tables();"

The most tests were updated: command "DROP SCHEMA arenadata_toolkit cascade;"
was added to the end of tests. It needs to keep clear environment after each
tests.

New tests were added: "tablespace_location" and "upgrade_test". Test
"tablespace_location" checks new field with different values: "pg_default",
"pg_global" and custom tablespace. Test "upgrade_test" checks updates from
version "1.0" to current.

Also, restarting cluster was added before running gpcheckcat tests.
It needs to minimize connection number for gpcheckcat.
gpcheckcat cleans all temp namespace (like "pg_temp_<CON_NUM>"
and "pg_toast_temp_<CON_NUM>"). gpcheckcat does not do it for active connections.
It may lead to the next situation: At the past (before previouse restarting
cluster) temporary table was created at pg_temp (while was connection with
number 467 for example). Next restarting cluster was executed. And when
gpcheckcat starts, it gets the same number for connection (467). gpcheckcat does
not clean pg_temp (and pg_toast_temp) because this connection is active. After
it gpcheckcat find error because namespace pg_temp still exist.

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61760

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/965102

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/965103

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/965101

RekGRpth commented on GNUmakefile.in:
```diff
@@ -208,7 +208,16 @@ installcheck-mirrorless:
+# it gets the same number for connection (467). gpcheckcat does not clean pg_temp
+# (and pg_toast_temp) because this connection is active. After it gpcheckcat find
+# error because namespace pg_temp still exist.
+	gpstop -ar && gpcheckcat -A
```
How does this change apply to arenadata_toolkit?

red1452 replied:
If you haven't any further questions, I will rebase PR to two commits (for arenadata_toolkit and for gpcheckcat).

RekGRpth replied:
may be fix gpcheckcat in separate patch?

RomaZe committed:
Upgrade Java 1.8 to Java 11 in Dockerfile (#674)

We are going to move our programs to be running on Java 11. All these programs
use the docker image (arenadata/Dockerfile) to run a lot of different kinds of
tests. So, we need Java 11 to test these programs. This patch replaces Java 1.8
with Java 11.

red1452 committed:
remove unnecessary changes

red1452 committed:
Merge branch 'adb-6.x-dev', remote-tracking branch 'origin' into ADBDEV-4598

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

red1452 committed:
Revert "Upgrade Java 1.8 to Java 11 in Dockerfile (#674)"

This reverts commit c097366900c7699f6e10fe5f29a763792d34af41.

RekGRpth commented on arenadata/Dockerfile:
```diff
@@ -12,9 +12,9 @@ RUN sed -i 's/\(override_install_langs*\)/# \1/' /etc/yum.conf && \
 RUN yum makecache && yum update -y ca-certificates && \
     rpm --import https://mirror.yandex.ru/centos/RPM-GPG-KEY-CentOS-7 && \
-    yum -y install epel-release java-1.8.0-openjdk-devel && \
+    yum -y install epel-release java-11-openjdk-devel && \
```
Is it change of this patch?

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63576

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046589

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046590

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046583

red1452 commented:
> PR description may be reformatted up to 80 chars so:

Done.

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,145 @@
+	DROP SCHEMA arenadata_toolkit CASCADE;
+
+-- Check create extension with the latest version after current was installed and dropped
+	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
+	IF from_version != '1.0'
+	THEN
+		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
+						UPDATE TO %1$I;$fmt$, from_version);
+	END IF;
+	PERFORM arenadata_toolkit.adb_create_tables();
+	PERFORM arenadata_toolkit.adb_collect_table_stats();
+	DROP EXTENSION arenadata_toolkit;
```
Why did not you removed this change from f90262d7ea86c4f3abae92ee180022699df16237?

red1452 replied:
Why should I do that? This code checks, that toolkit will be installed correctly, even if there are some relations from previous installation.

RekGRpth replied:
You removed all from that commit except pointed code

red1452 replied:
This code is not related to code, which removed temporary tables.

RekGRpth replied:
What does this code check? And why was it added with said commit?

red1452 replied:
> What does this code check?

After install and remove toolkit, there are some relations, which was created by toolkit's functions "adb_create_tables" and "adb_collect_table_stats". These relations are not removed automaticly. And this code checks, that next installation will be success.

> And why was it added with said commit?

Maybe, I forgot to add short comment to commit message about this code.

RekGRpth replied:
> there are some relations, which was created by toolkit's functions "adb_create_tables" and "adb_collect_table_stats"

But aren’t they deleted above up when the schema is deleted in cascade?

RekGRpth replied:
> this code checks, that next installation will be success.

Doesn't the code below check this?

bimboterminator1 replied:
> And this code checks, that next installation will be success.

red1452 is talking about code below

red1452 replied:
This part is preparation for testing:
```sql
	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
	IF from_version != '1.0'
	THEN
		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
						UPDATE TO %1$I;$fmt$, from_version);
	END IF;
	PERFORM arenadata_toolkit.adb_create_tables();
	PERFORM arenadata_toolkit.adb_collect_table_stats();
	DROP EXTENSION arenadata_toolkit;
```

This code checks what I said:
```sql
	CREATE EXTENSION arenadata_toolkit;
	PERFORM arenadata_toolkit.adb_create_tables();
	PERFORM arenadata_toolkit.adb_collect_table_stats();

-- Check the result
	IF (SELECT default_version = installed_version
		FROM pg_available_extensions
		WHERE name='arenadata_toolkit')
	THEN
		RETURN NEXT from_version || ': create the latest check';
	END IF;

```

red1452 replied:
> But aren’t they deleted above up when the schema is deleted in cascade?

This code checks the situation, when command "DROP SCHEMA arenadata_toolkit CASCADE;" is not executed.

RekGRpth replied:
> This part is preparation for testing:
> 
> ```sql
> 	CREATE EXTENSION arenadata_toolkit VERSION '1.0';
> 	IF from_version != '1.0'
> 	THEN
> 		EXECUTE FORMAT($fmt$ALTER EXTENSION arenadata_toolkit
> 						UPDATE TO %1$I;$fmt$, from_version);
> 	END IF;
> 	PERFORM arenadata_toolkit.adb_create_tables();
> 	PERFORM arenadata_toolkit.adb_collect_table_stats();
> 	DROP EXTENSION arenadata_toolkit;
> ```

Why wasn't this needed before the commit https://github.com/arenadata/gpdb/commit/f90262d7ea86c4f3abae92ee180022699df16237?

RekGRpth replied:
> This code checks the situation, when command "DROP SCHEMA arenadata_toolkit CASCADE;" is not executed.

OK, I think you should remove lines 106-107 and 109-117, as it was before commit f90262d.

red1452 replied:
> OK, I think you should remove lines 106-107 and 109-117, as it was before commit [f90262d](https://github.com/arenadata/gpdb/commit/f90262d7ea86c4f3abae92ee180022699df16237).

NO, It will change test's logic: now this part of test checks creating the latest version of extension after each versions . If I remove these lines, test will check creating the latest version of extension after the latest version.

red1452 replied:
> Why wasn't this needed before the commit [f90262d](https://github.com/arenadata/gpdb/commit/f90262d7ea86c4f3abae92ee180022699df16237)?

At first redaction of this patch (where was to arguments at function `do_upgrade_test_for_arenadata_toolkit`: `from_version` and `to_version`) `to_version` was different for some executions, but at the last redaction of this patch `to_version` is the last version of toolkit and does not change at the test.

RekGRpth replied:
I still think you should remove lines 106-107 and 109-117, because they are redundant

red1452 replied:
Let's compare two variants of test:
At line 67, arenadata_toolkit is updated to the latest version. At line 105-106 areanadata_toolkit is dropped and all relations are removed. At lines 110-114 arenadata_toolkit is created with version `from_version` (it is important). All waste will be created at version `from_version` of arenadata_toolkit. If I remove lines 106-107 and 109-117 than for each execution of `do_upgrade_test_for_arenadata_toolkit` check at lines 119-129 will do the same things, because waste will be from the latest version of arenadata_toolkit.

RekGRpth replied:
OK, this code is needed for test create after dropping old version?
old -> create tables -> collect stats -> drop -> create extension 
But where are same tests for other cases
1) old -> drop -> create extension
2) old -> create tables -> drop -> create extension
?

red1452 replied:
> But where are same tests for other cases

I think, It can not check more than my case (old -> create tables -> collect stats -> drop -> create extension), but the test will be done longer without any helpful result. Now, this test is the longest test at arenadata_toolkit.

bimboterminator1 replied:
>old -> drop -> create extension
old -> create tables -> drop -> create extension

We definitely should rework the upgrade tests framework in future. I bet we may leave here only tests related to 1.4 specification. Other general cases like above would be introduced when we add an extensible version of upgrade tests. What do you think?

RekGRpth replied:
Previously, the comment on line 108 referred to the code on lines 119-121, but now this code is left without a comment.

bimboterminator1 commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,145 @@
+	IF EXISTS (SELECT 1
+			   FROM pg_tables
+			   WHERE schemaname='arenadata_toolkit' AND
+			         tablename SIMILAR TO 'db_files_history_backup_[0-9]{8}t[0-9]{6}')
```
What will be when we update extension to 1.5? 
We will execute do_upgrade_test for from_version 1.4(otherwise what's the point of the whole upgrade test), 
and what will be with that check? We just won't print it and it will be expected result?

red1452 replied:
At this case backup is not needed (field `tablespace_location` has already existed at table `db_files_history`). When we execute `SELECT do_upgrade_test_for_arenadata_toolkit('1.4') ORDER BY 1;`, we will not show line : "1.4:  db_files_history_backup check". I think, it is normal case, we can check, that extra actions are not done.

bimboterminator1 commented:
Add the header to the patch description, according to the paragraph 2.2 of description guidelines.

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,145 @@
+END$$
+LANGUAGE plpgsql;
+
+SELECT do_upgrade_test_for_arenadata_toolkit('1.0') ORDER BY 1;
+SELECT do_upgrade_test_for_arenadata_toolkit('1.1') ORDER BY 1;
+SELECT do_upgrade_test_for_arenadata_toolkit('1.2') ORDER BY 1;
+SELECT do_upgrade_test_for_arenadata_toolkit('1.3') ORDER BY 1;
```
```suggestion
create external web table ver (x text) execute E'find $(pg_config --sharedir) -name "arenadata_toolkit--*--*.sql" -type f -printf "%f\\n" | grep -oP "(\\d+\\.\\d+)\\.sql" | grep -oP "(\\d+\\.\\d+)"' on master format 'text';
SELECT do_upgrade_test_for_arenadata_toolkit(x) from ver ORDER BY 1;
```

red1452 committed:
optimization of upgrade test

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/upgrade_test.sql:
```diff
@@ -0,0 +1,149 @@
+END$$
+LANGUAGE plpgsql;
+
+create external web table ver (x text)
```
1) use upper case
2) drop table after use
3) name table and column more understandable

red1452 committed:
small fixes

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053462

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053463

RekGRpth commented:
> Test upgrade_test checks updates from version
> 1.0 to current.

May be "from all previous versions to current"?

red1452 commented:
> May be "from all previous versions to current"?

Updated.

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053452

andr-sokolov commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,375 @@
+AS $$
+BEGIN
+	IF NOT EXISTS (
+		SELECT 1
```
when the exists clause is used, then select's target list does not matter. So remove please '1', use empty target list.

red1452 committed:
small optimizations

RekGRpth commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.3--1.4.sql:
```diff
@@ -0,0 +1,357 @@
+
+DO $$
+BEGIN
+	IF (SELECT a.attrelid IS NULL
```
You can do some more optimization here and in other similar places like this:
```suggestion
	IF a.attrelid IS NULL
```

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64137

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1059226

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1059227

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1059218

red1452 committed:
fixes

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64191

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1060881

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1060882

RekGRpth commented on gpcontrib/arenadata_toolkit/.gitignore:
```diff
@@ -1,4 +1,6 @@
 # Generated subdirectories and files
+/expected/tablespace_location.out
```
?

red1452 replied:
This file is generated from new file: "output/tablespace_location.out"

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1060873

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1061510

RekGRpth commented on gpcontrib/arenadata_toolkit/input/tablespace_location.source:
```diff
@@ -0,0 +1,70 @@
+SELECT arenadata_toolkit.adb_create_tables();
+RESET client_min_messages;
+
+\! rm -rf @testtablespace@ && mkdir @testtablespace@;
```
Core regress never delete nor create @testtablespace@, instead it keeps folders in own directory.

red1452 replied:
For arenadata_toolkit this is specific path. New directory is created at arenadata_toolkit sources. This commands are needed to keep clean this directory for tests. Also at other regress tests there are creating and removing commands: `regress/input/gp_check_files.source`, `regress/input/gp_tablespace.source`.

RekGRpth replied:
> at other regress tests there are creating and removing commands: `regress/input/gp_check_files.source`, `regress/input/gp_tablespace.source`.

No! Only subdirectories

RekGRpth replied:
> this is specific path. New directory is created at arenadata_toolkit sources

So I suggest creating a directory, like in the core regress.

red1452 replied:
It does not work:
```
\! rm -rf /home/evgeniy/gpdb/adb-6.x-gpdb/gpcontrib/arenadata_toolkit/testtablespace/*;
 CREATE TABLESPACE test_tblspc LOCATION '/home/evgeniy/gpdb/adb-6.x-gpdb/gpcontrib/arenadata_toolkit/testtablespace';
+ERROR:  directory "/home/evgeniy/gpdb/adb-6.x-gpdb/gpcontrib/arenadata_toolkit/testtablespace" does not exist
```

RekGRpth replied:
> It does not work:

because you should create directory in repo first, like in core regress

red1452 replied:
For other tests, the same commands is executed at `regress/GNUmakefile`:
```
.PHONY: tablespace-setup
tablespace-setup:
	rm -rf ./testtablespace ./testtablespace_*
	mkdir -p ./testtablespace ./testtablespace_otherloc ./testtablespace_unlogged \
	./testtablespace_existing_version_dir/1/GPDB_99_399999991/ \
...
```
What is difference between doing these commands at the test and doing these commans at makefile?

RekGRpth replied:
see `ls src/test/regress/testtablespace*`

red1452 replied:
The main idea of using `@testtablespace@` in the test: not to write path of the directory for the test. If I do as you suggest, I will write this path at two commands. What is the point to use this type of test, if path will be written manually?

RekGRpth replied:
> I will write this path at two commands

In which two commands? You only need to manually add the directory to the repository once (as is done in the core regress).

red1452 replied:
Do you understand the main idea of using `@testtablespace@` at the test? But you suggest to print this path at the code. Do you think about situation, if at `pg_regress` this directory will be changed? Our test will break. And we must change this path. My solution allow us to avoid the problem.

RekGRpth replied:
> But you suggest to print this path at the code

No, no! Only

> manually add the directory to the repository once (as is done in the core regress).

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64467

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067930

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067931

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067925

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4598

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64568

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1071064

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1071065
